---
title: "ERM & Blood Analysis -- Report 1"
author: "Yiling Zhang"
date: "`r Sys.Date()`"
output: 
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
    df_print: paged
    toc: true
toc-title: "Table of Contents"
vignette: >
  %\VignetteIndexEntry{ERM & Blood Analysis -- Report 1}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
	message = FALSE,
	warning = FALSE,
	fig.align = "center",
  fig.width = 8
)
```

# 1. Initialize Data
```{r}
library(Seurat)
library(dplyr)
library(ggplot2)
```

## 1.1 Read in and Initialize Seurat Obejct
```{r}
raw.p1ERM <- read.table("../data/rawMatix/patient_1/PDR-ERM-0426_matrix.tsv",header = T) %>% CreateSeuratObject(project = "p1-ERM")
raw.p1Blood <- read.table("../data/rawMatix/patient_1/PDR-Blood-0426_matrix.tsv", header = T) %>% CreateSeuratObject(project = "p1-Blood")

raw.p2ERM <- read.table("../data/rawMatix/patient_2/PDR-ERM_matrix.tsv", header = T) %>% CreateSeuratObject(project = "p2-ERM")
raw.p2Blood <- read.table("../data/rawMatix/patient_2/PDR-Blood_matrix.tsv", header = T) %>% CreateSeuratObject(project = "p2-Blood")
```


## 1.2 Merge data
```{r}
mergeData <- merge(x = raw.p1ERM, y = c(raw.p1Blood, raw.p2ERM, raw.p2Blood))
```

```{r include=FALSE}
rm(raw.p1ERM, raw.p1Blood, raw.p2ERM, raw.p2Blood)
```

## 1.3 Check data
```{r}
mergeData
```

```{r}
mergeData@assays$RNA[1:5,1:5]
```

```{r}
mergeData@meta.data[1:5,]
```

```{r}
mergeData@meta.data %>% 
  	ggplot(aes(x=orig.ident, fill=orig.ident)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells")
```

# 2. QC
## 2.1 Data examining
cell number:    
```{r}
dim(mergeData)
```


Counts per cell:    
```{r}
counts_per_cell <- Matrix::colSums(mergeData)
hist(log10(counts_per_cell+1),main='counts per cell',col='Thistle')
```


Genes per cell:    
```{r}
genes_per_cell <- Matrix::colSums(mergeData@assays$RNA@counts > 0)
hist(log10(genes_per_cell+1), main='genes per cell', col='Thistle')
```

## 2.2 Remove Doublets

* [DoubletDecon](https://github.com/EDePasquale/DoubletDecon)  
Fail to install


* [DoubletFinder](https://github.com/chris-mcginnis-ucsf/DoubletFinder)

```{r}
library(DoubletFinder)
library(patchwork)
library(clustree)
```

### 2.2.1 Pre-process Seurat object
```{r message=FALSE, warning=FALSE}
mergeData <- SCTransform(mergeData, verbose = F)
mergeData <- RunPCA(mergeData, npcs = 30, verbose = F)
ElbowPlot(mergeData,ndims = 30)
```

```{r}
# Set dim 1:20
pc.num=1:20 
```

```{r}
mergeData <- RunUMAP(mergeData, dims=pc.num,verbose = F)
mergeData <- FindNeighbors(mergeData, reduction = "pca", dims = pc.num)
```

```{r fig.height=8}
checkRes <- FindClusters(object = mergeData, resolution = c(seq(.1,1.6,.2)), verbose = F)
clustree(checkRes@meta.data,prefix = "SCT_snn_res.")
```

```{r}
mergeData <- FindClusters(mergeData, resolution = 0.7)
```

### 2.2.2 pK Identification

```{r eval=FALSE}
# Optimize the parameters
sweep.res.list <- paramSweep_v3(mergeData, PCs = pc.num, sct = T)
# Use log transform
sweep.stats <- summarizeSweep(sweep.res.list, GT = F)
```

```{r include=FALSE}
load("../data/tmp.data/sweepStats.RData")
```

```{r}
# Show the best parameter
bcmvn <- find.pK(sweep.stats)
```

```{r}
# Extract the best pK
pK_bcmvn <- bcmvn$pK[which.max(bcmvn$BCmetric)] %>% as.character() %>% as.numeric()
pK_bcmvn
```

### 2.2.3 Homotypic Doublet Proportion Estimate

>doublet rate: https://singleronbio.com/product/?type=detail&id=8

```{r}
# Estimate the percentage of homotypic doublets
homotypic.prop <- modelHomotypic(mergeData$seurat_clusters)

DoubletRate = 0.0268
nExp_poi <- round(DoubletRate*ncol(mergeData)) 
# Adjust for homotypic doublets
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
```

### 2.2.4 Run DoubletFinder

```{r eval=FALSE}
mergeData <- doubletFinder_v3(mergeData, PCs = pc.num, pN = 0.25, pK = pK_bcmvn, 
                             nExp = nExp_poi.adj, reuse.pANN = F, sct = T)
```

```{r include=FALSE, eval=TRUE}
mergeData <- readRDS("../data/SeuratObj/merged.RDS")
```

```{r}
# Present the res, classification info is saved in meta.data
DF.name <- colnames(mergeData@meta.data)[grepl("DF.classification", colnames(mergeData@meta.data))]
# Visualization
cowplot::plot_grid(ncol = 2, DimPlot(mergeData, group.by = "orig.ident") + NoAxes(), 
                   DimPlot(mergeData, group.by = DF.name) + NoAxes())
```


```{r}
DF.name <- colnames(mergeData@meta.data)[grepl("DF.classification", colnames(mergeData@meta.data))]
mergeData.singlet <- mergeData[, mergeData@meta.data[, DF.name] == "Singlet"]
mergeData.singlet
```


## 2.3 Calculate QC
### 2.3.1 Generating quality metrics

**mitochondrial ratio:**
```{r}
mergeData.singlet[["percent.mt"]] <- PercentageFeatureSet(object = mergeData.singlet, pattern = "^MT-", assay = "RNA")
mergeData.singlet$percent.mt <- mergeData.singlet@meta.data$percent.mt / 100
```

**ribosome ratio:** 
```{r}
mergeData.singlet[["percent.ribo"]] <- PercentageFeatureSet(object = mergeData.singlet, pattern = "^RP[SL]", assay = "RNA")
mergeData.singlet$percent.ribo <- mergeData.singlet@meta.data$percent.ribo / 100
```

**hemoglobin ratio:**
```{r}
mergeData.singlet[["percent.hb"]] <- PercentageFeatureSet(object = mergeData.singlet, pattern = "^HB[^(P)]", assay = "RNA")
mergeData.singlet$percent.ribo <- mergeData.singlet@meta.data$percent.ribo / 100
```

**number of genes detected per UMI:**    
represent the complexity of our dataset (more genes detected per UMI, more complex our data)
```{r}
mergeData.singlet$log10GenesPerUMI <- log10(mergeData.singlet$nFeature_RNA) / log10(mergeData.singlet$nCount_RNA)
```

### 2.3.2 Assessing the quality metrics:   

**UMI counts (transcripts) per cell**  
The UMI counts per cell should generally be above 500, that is the low end of what we expect. If UMI counts are between 500-1000 counts, it is usable but the cells probably should have been sequenced more deeply.    

* Visualize the number UMIs/transcripts per cell    
```{r}
mergeData.singlet@meta.data %>% 
  	ggplot(aes(color=orig.ident, x=nCount_RNA, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)
```

**Genes detected per cell**  
For high quality data, the proportional histogram should contain a single large peak that represents cells that were encapsulated. 

* Visualize the distribution of genes detected per cell via histogram
```{r}
mergeData.singlet@meta.data %>% 
  	ggplot(aes(color=orig.ident, x=nFeature_RNA, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)
```


**UMIs vs. genes detected**  
Cells that are poor quality are likely to have low genes and UMIs per cell, and correspond to the data points in the bottom left quadrant of the plot. Good cells will generally exhibit both higher number of genes per cell and higher numbers of UMIs. Cells have a high number of UMIs but only a few number of genes could be dying cells, but also could represent a population of a low complexity celltype (i.e red blood cells).  


Mitochondrial read fractions are only high in particularly low count cells with few detected genes. This could be indicative of damaged/dying cells whose cytoplasmic mRNA has leaked out through a broken membrane, and thus, only mRNA located in the mitochondria is still conserved.   

* Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
```{r}
mergeData.singlet@meta.data %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~orig.ident)
```

**Complexity**
Outlier cells in these samples might be cells that have a less complex RNA species than other cells. Sometimes we can detect contamination with low complexity cell types like red blood cells via this metric. Generally, we expect the novelty score to be above 0.80.

* Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
```{r}
mergeData.singlet@meta.data %>%
  	ggplot(aes(x=log10GenesPerUMI, color = orig.ident, fill=orig.ident)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = 0.8)
```

**Mitochondrial counts ratio**
This metric can identify whether there is a large amount of mitochondrial contamination from dead or dying cells.

* Visualize the distribution of mitochondrial gene expression detected per cell
```{r}
mergeData.singlet@meta.data %>% 
  	ggplot(aes(color=orig.ident, x=percent.mt, fill=orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)
```


**Ribosome counts ratio**
This metric can identify whether there is a large amount of ribosome contamination from dead or dying cells.

* Visualize the distribution of ribosome gene expression detected per cell
```{r}
mergeData.singlet@meta.data %>% 
  	ggplot(aes(color=orig.ident, x=percent.ribo, fill=orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.01)
```


**Hemoglobin counts ratio**
This metric can identify whether there is a large amount of hemoglobin contamination from red blood cell leak.

* Visualize the distribution of mitochondrial gene expression detected per cell
```{r}
mergeData.singlet@meta.data %>% 
  	ggplot(aes(color=orig.ident, x=percent.hb, fill=orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.25)
```


## 2.4 Filtering
### 2.4.1 Cell-level filtering
* nCount > 500  
* nFeature > 250  
* log10GenesPerUMI > 0.8  
* percent.mt < 0.2  
* percent.ribo < 0.01  
* percent.hb < 0.25  

```{r}
filteredData <- subset(x = mergeData.singlet, 
                         subset= (nCount_RNA >= 500) & 
                           (nFeature_RNA >= 250) & 
                           (log10GenesPerUMI > 0.80) & 
                           (percent.mt < 0.20) & 
                           (percent.ribo < 0.01) & 
                           (percent.hb < 0.25))
```

### 2.4.2 Gene-level filtering
```{r eval=FALSE}
par(mar = c(4, 8, 2, 1))
C <- filteredData@assays$RNA@counts
C <- Matrix::t(Matrix::t(C)/Matrix::colSums(C)) * 100
most_expressed <- order(apply(C, 1, median), decreasing = T)[20:1]
boxplot(as.matrix(t(C[most_expressed,])), cex = 0.1, las = 1, xlab = "% total count per cell", col = (scales::hue_pal())(20)[20:1], horizontal = TRUE)
```

First we will remove genes that have zero expression in all cells. Additionally, we will perform some filtering by prevalence. If a gene is only expressed in a handful of cells, it is not particularly meaningful. For our data we choose to **keep only genes which are expressed in 5 or more cells**.

```{r}
# Extract counts
counts <- GetAssayData(object = filteredData, slot = "counts",assay = "RNA")

# Output a logical vector for every gene on whether the more than zero counts per cell
nonzero <- counts > 0

# Sums all TRUE values and returns TRUE if more than 5 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 5

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]

# Reassign to filtered Seurat object
filteredData2 <- CreateSeuratObject(filtered_counts, meta.data = filteredData@meta.data)
filteredData2
```

# 3. Integration
## 3.1 Cell cycle scoring check
```{r}
cell_cycle_genes <- read.csv("../data/cell-cycle/Homo_sapiens.csv")
head(cell_cycle_genes)
```

ID transfer via annotation databases
```{r}
library(AnnotationHub)
library(ensembldb)
```

* Query Ensembl 
```{r eval=FALSE}
# Connect to AnnotationHub
ah <- AnnotationHub()

# Access the Ensembl database for organism
ahDb <- query(ah, 
              pattern = c("Homo sapiens", "EnsDb"), 
              ignore.case = TRUE)

# Acquire the latest annotation files
id <- ahDb %>%
        mcols() %>%
        rownames() %>%
        tail(n = 1)

# Download the appropriate Ensembldb database
edb <- ah[[id]]

# Extract gene-level information from database
annotations <- genes(edb, 
                     return.type = "data.frame")

# Select annotations of interest
annotations <- annotations %>%
        dplyr::select(gene_id, gene_name, seq_name, gene_biotype, description)
```

```{r eval=TRUE, include=FALSE}
load("../data/annotation/annotations.RData")
```

* Annotations to get the corresponding gene names for the Ensembl IDs of the cell cycle genes
```{r}
# Get gene names for Ensembl IDs for each gene
cell_cycle_markers <- dplyr::left_join(cell_cycle_genes, annotations, by = c("geneID" = "gene_id"))

# Acquire the S phase genes
s_genes <- cell_cycle_markers %>%
        dplyr::filter(phase == "S") %>%
        pull("gene_name")
        
# Acquire the G2M phase genes        
g2m_genes <- cell_cycle_markers %>%
        dplyr::filter(phase == "G2/M") %>%
        pull("gene_name")
```

## 3.2 SCTransform

```{r}
split_seurat <- SplitObject(filteredData2, split.by = "orig.ident")
split_seurat
```

Perform the cell cycle scoring and sctransform on all samples.
```{r}
for (i in 1:length(split_seurat)) {
    split_seurat[[i]] <- NormalizeData(split_seurat[[i]], verbose = F)
    split_seurat[[i]] <- CellCycleScoring(split_seurat[[i]], g2m.features=g2m_genes, s.features=s_genes)
    split_seurat[[i]] <- SCTransform(split_seurat[[i]], vars.to.regress = c("percent.mt"), variable.features.n = 5000,  verbose = F, seed.use = 27)
}
```

## 3.2 Integration

First, we need to specify that we want to use 3000 most variable genes identified by SCTransform for the integration. By default, this function only selects the top 2000 genes.
```{r eval=FALSE}
# Select the most variable features to use for integration
integ_features <- SelectIntegrationFeatures(object.list = split_seurat, nfeatures = 3000) 
```

Then, we need to prepare the SCTransform object for integration.
```{r eval=FALSE}
# Prepare the SCT list object for integration
split_seurat <- PrepSCTIntegration(object.list = split_seurat, 
                                   anchor.features = integ_features)
```

Now, we are going to perform CCA, find the best buddies or anchors and filter incorrect anchors. 
```{r eval=FALSE}
# Find best buddies - can take a while to run
integ_anchors <- FindIntegrationAnchors(object.list = split_seurat, 
                                        normalization.method = "SCT", 
                                        anchor.features = integ_features)
```

Finally, we can integrate across conditions.
```{r eval=FALSE}
# Integrate across conditions
integrated <- IntegrateData(anchorset = integ_anchors, normalization.method = "SCT")
integrated
```

```{r include=FALSE}
integrated <- readRDS("../data/SeuratObj/integrated.rds")
```

# 4. Dimensional reduction
## 4.1 PCA
```{r}
# Run PCA
integrated <- RunPCA(object = integrated, npcs = 30, verbose = F)
ElbowPlot(integrated, ndims = 30)
```

```{r fig.width=12}
pc.num=1:25

# Plot PCA
PCAPlot(integrated, split.by = "orig.ident")  
```

## 4.2 UMAP
```{r}
# Run UMAP
integrated <- RunUMAP(integrated, dims = pc.num, reduction = "pca")

# Plot UMAP                             
DimPlot(integrated, group.by = "orig.ident") 
```

## 4.2 tSNE
```{r}
# Run tSNE
integrated <- RunTSNE(object = integrated, dims.use = pc.num)

# Plot tSNE
DimPlot(object = integrated, group.by = "orig.ident", reduction = "tsne")
```

```{r fig.width=12}
FeaturePlot(object = integrated, features = c("GPNMB"), cols = c("grey", "blue"), reduction = "tsne",split.by = "orig.ident")
```

```{r fig.width=12}
FeaturePlot(object = integrated, features = c("CD3D"), cols = c("grey", "blue"), reduction = "tsne",split.by = "orig.ident")
```


```{r fig.width=12,fig.height=7}
FeaturePlot(object = integrated, features = c("CD8A", "CD8B", "ZNF683"), cols = c("grey", "blue"), reduction = "tsne", split.by = "orig.ident")
```

# 5. Clustering cells based on top PCs
## 5.1 Identify significant PCs
```{r}
# Plot the elbow plot
ElbowPlot(object = integrated, ndims = 30)
```

## 5.2 Cluster the cells
```{r}
# Determine the K-nearest neighbor graph
integrated <- FindNeighbors(object = integrated, dims = pc.num)
```

```{r fig.height=8}
checkRes <- FindClusters(object = integrated, resolution = c(seq(.1,1.6,.2)), verbose = F)
clustree(checkRes@meta.data,prefix = "integrated_snn_res.")
```

```{r}
integrated <- FindClusters(integrated, resolution = 0.7)
```

```{r}
# Plot the UMAP
DimPlot(integrated, reduction = "umap", label = TRUE, label.size = 6)
```

```{r fig.width=12}
# Plot the UMAP
DimPlot(integrated, reduction = "umap", label = TRUE, label.size = 6, split.by = "orig.ident")
```